docs/design/2021-08-23-history-hot-regions.md

```diff
@@ -0,0 +1,269 @@
+  | PEER_ID     | bigint(21)  | YES  |      | NULL    |       | // new
+  | IS_LEARNER  | tinyint(1)  | NO   |      | 0       |       | // new
+  | IS_LEADER   | tinyint(1)  | NO   |      | 0       |       | // new
+  | TYPE        | varchar(64) | YES  |      | NULL    |       |
+  | HOT_DEGREE  | bigint(21)  | YES  |      | NULL    |       | // rename max_hot_degree to HOT_DEGREE
+  | FLOW_BYTES  | double      | YES  |      | NULL    |       |
+  | KEY_RATE    | double      | YES  |      | NULL    |       | // new
+  | QUERY_RATE  | double      | YES  |      | NULL    |       | // new
+  +-------------+-------------+------+------+---------+-------+
+  | REGION_COUNT| bigint(21)  | YES  |      | NULL    |       | // deleted fields
+  ```
+  * Add `UPDATE_TIME` to support history.
+  * Add `STORE_ID` ,  `PEER_ID`,  `IS_LAEDER`to track the machine of region.
+  * Rename `MAX_HOT_DEGREE`  to `HOT_DEGREE` for precise meaning in history scenario.
+  * Add `IS_LEARNER` and `IS_LEADER` to show role of regions.
+  * Add `KEY_RATE` and `QUERY_RATE` for future expansion in hotspot determination dimensions.
+  * Remove `REGION_COUNT` for disuse and repeat with `STORE_ID`.
+
+2. Data size estimation 
+
+    one record size: 1 * 4B(timestamp) + 1*1B(tinyint) + 6 * 8B(bitint) + 3 * 8B(double)  + 4 * 64B(varchar(64)) = 333B
+    Below table show data size per day and per month in 5,10,15 minutes record interval respectively given the maximum number of hotspot regions  is 1000:
+
+    | Record Length (B) | Time Interval (Min) | Data Size Per Day (MB) | Data Size Per Month (MB) |
+    | ----------------- | ------------------- | ---------------------- | ------------------------ |
+    | 333               | 5                   | 91.46118164            | 2743.835449              |
+    | 333               | 10                  | 45.73059082            | 1371.917725              |
+    | 333               | 15                  | 30.48706055            | 914.6118164              |
+
+### PD
+1. Timing write：
+
+     The leader of PD will periodically encrypt  `START_KEY` and `END_KEY`  in data,and write hotspot region data into `LevelDB`.The write interval can be configured. The write fieldes are: `REGION_ID`, `TYPE`,  `HOT_DEGREE`, `FLOW_BYTES`, `KEY_RATE`, `QUERY_RATE`, `STORE_ID`, `PEER_ID`, `UPDATE_TIME`, `START_KEY`,  `END_KEY`.
+
+2. Timing delete
+
+   PD runs the delete check task periodically, and deletes the hotspot region data that exceeds the configured TTL time.
+
+3. Pull interface
+
+   PD query the data in `LevleDB` with filters pushed down, decodes the data and returns to TiDB.
+
+4. New options
+
+   There are two config options need to be add in PD’s `config.go`:
+
+   * `HisHotRegionSaveInterval`:  time interval for pd to record hotspot region information, default: 10 minutes.
+   * `HisHotRegionTTL`: maximum hold day for his hot region, default: 7 days. 0 means close.
+   
+5. GC
+
+     The amount of data stored for one month is as follows.
```

Comment:
Some theoretical formulas can be added. For example, there are at most 1000 hotspot information, and each hotspot information is about how many bytes.
example: 
one month data write = 30(days) * 24(hours) * 6 (items/hours) * 100 B (each item size) * 1000 (max item numbers) = 411 MB. and the default save TTL is 7 days/